packages/fdc3-agent-proxy/src/listeners/AbstractListener.ts

```diff
@@ -1,54 +1,115 @@
   readonly messaging: Messaging;
-  readonly handler: X;
+  private readonly subscribeRequestType: Y['type'];
+  private readonly subscribeResponseType: SubscriptionResponse['type'];
+  private readonly unsubscribeRequestType: UnsubscribeRequest['type'];
+  private readonly unsubscribeResponseType: UnsubscribeResponse['type'];
+  private readonly subscriptionPayload: Y['payload'];
   id: string | null = null;
+  readonly handler: X;
 
   constructor(
     messaging: Messaging,
-    private readonly payloadDetails: Record<string, string | null>,
+    subscriptionPayload: Y['payload'],
     handler: X,
-    private readonly subscribeType: string,
-    private readonly unsubscribeType: string
+    subscribeRequestType: Y['type'],
+    subscribeResponseType: SubscriptionResponse['type'],
+    unsubscribeRequestType: UnsubscribeRequest['type'],
+    unsubscribeResponseType: UnsubscribeResponse['type']
   ) {
     this.messaging = messaging;
     this.handler = handler;
+    this.subscriptionPayload = subscriptionPayload;
+    this.subscribeRequestType = subscribeRequestType;
+    this.subscribeResponseType = subscribeResponseType;
+    this.unsubscribeRequestType = unsubscribeRequestType;
+    this.unsubscribeResponseType = unsubscribeResponseType;
   }
 
-  abstract filter(m: any): boolean;
+  abstract filter(m: AgentEventMessage): boolean;
 
-  abstract action(m: any): void;
+  abstract action(m: AgentEventMessage): void;
+
+  async unsubscribe(): Promise<void> {
+    if (this.id) {
+      this.messaging.unregister(this.id!);
+      const unsubscribeMessage: UnsubscribeRequest = {
+        meta: this.messaging.createMeta(),
+        payload: {
+          listenerUUID: this.id,
+        },
+        type: this.unsubscribeRequestType,
+      };
+      await this.messaging.exchange<UnsubscribeResponse>(unsubscribeMessage, this.unsubscribeResponseType);
+      return;
+    } else {
+      console.error("This listener doesn't have an id and hence can't be removed!");
+    }
+  }
 
-  async listenerNotification(type: string): Promise<string | null> {
-    const requestType = `${type}Request`;
-    const responseType = `${type}Response`;
-    const notificationMessage = {
+  async register(): Promise<void> {
+    const subscribeMessage: Y = {
       meta: this.messaging.createMeta(),
-      payload: this.id ? { listenerUUID: this.id } : { ...this.payloadDetails },
-      type: requestType,
-    };
+      payload: this.subscriptionPayload,
+      type: this.subscribeRequestType,
+    } as Y;
 
-    const response = await this.messaging.exchange<any>(notificationMessage, responseType);
-    return response?.payload?.listenerUUID ?? null;
-  }
+    const response = await this.messaging.exchange<SubscriptionResponse>(subscribeMessage, this.subscribeResponseType);
+    this.id = response?.payload?.listenerUUID ?? null;
 
-  async unsubscribe(): Promise<void> {
     if (!this.id) {
```

Comment:
By 'here' do you mean in `register()` or in `unsubscribe()`? I'm going to assume `register()`. 

The problem is that we're integrating with 3rd party code that we don't control. We can totally get here if they send an invalid message that has neither the `listenerUUID` nor `error` properties set in the payload. Yes, we could move to validating all messages as they arrive and throw in exchange, but that will come with a performance hit and we'd need a new error defined in the API to return to the user in every API call. Whereas, with this pattern we're validating just the field we're dependent on and providing feedback to the DA developer (or user who can pass it on to them) about what they did wrong and what will fail because of it. Having spent a lot of time debugging interop code other people wrote, I'm keen on error messages like this - more than I am worried about code coverage for the log line (which is hard to achieve).

Reply:
yeah I think we came up with a good approach for this a few comments back - let's use it here too

Reply:
👍